[WTF] Ensure isMainThread semantics for Linux match Apple/BSD semantics

jameshilliard opened the pull request:
#### 50d7bfe855ca043cc72ec09b3d3d75100015101b
<pre>
[WTF] Ensure isMainThread semantics for Linux match Apple/BSD semantics
<a href="https://bugs.webkit.org/show_bug.cgi?id=242690">https://bugs.webkit.org/show_bug.cgi?id=242690</a>

Reviewed by Darin Adler.

It&apos;s not entirely clear if the current implementation semantics are fully
identical between Linux and Apple/BSD systems. The Apple/BSD pthread_main_np
function has 3 potential return values, 1 if the thread is the initial thread,
0 if it is not or -1 if thread is not initialized/unknown, lets make sure we
only return true for isMainThread if the return for pthread_main_np is 1.

On linux checking if we are on the initial thread can be accomplished by
checking the equality of getpid and gettid. Note we use a syscall for calling
gettid as the gettid() libc function may not be guaranteed to be supported by
all libc&apos;s on Linux unlike the syscall(which is also used elsewhere in WebKit).

* Source/WTF/wtf/generic/MainThreadGeneric.cpp:
(WTF::initializeMainThreadPlatform):
(WTF::isMainThread):

Canonical link: <a href="https://commits.webkit.org/252485@main">https://commits.webkit.org/252485@main</a>
</pre>

darinadler commented on Source/WTF/wtf/generic/MainThreadGeneric.cpp:
```diff
@@ -31,30 +31,39 @@
 {
 #if HAVE(PTHREAD_MAIN_NP)
-    return pthread_main_np();
+    return pthread_main_np() == 1;
```
The rest of this patch changes behavior on Linux. This part attempts to fix a bug on the `HAVE(PTHREAD_MAIN_NP)` platforms. I’d like to understand how we’re testing this part.

jameshilliard replied:
I guess we can do something like this to make sure this call isn't failing:
```cpp
    int mainThreadNp = pthread_main_np();
    ASSERT(mainThreadNp != -1);
    return mainThreadNp == 1;
```

jameshilliard committed:
[WTF] Ensure isMainThread semantics for Linux match Apple/BSD semantics
https://bugs.webkit.org/show_bug.cgi?id=242690

Reviewed by Darin Adler.

It's not entirely clear if the current implementation semantics are fully
identical between Linux and Apple/BSD systems. The Apple/BSD pthread_main_np
function has 3 potential return values, 1 if the thread is the initial thread,
0 if it is not or -1 if thread is not initialized/unknown, lets make sure we
only return true for isMainThread if the return for pthread_main_np is 1.

On linux checking if we are on the initial thread can be accomplished by
checking the equality of getpid and gettid. Note we use a syscall for calling
gettid as the gettid() libc function may not be guaranteed to be supported by
all libc's on Linux unlike the syscall(which is also used elsewhere in WebKit).

* Source/WTF/wtf/generic/MainThreadGeneric.cpp:
(WTF::initializeMainThreadPlatform):
(WTF::isMainThread):

Canonical link: https://commits.webkit.org/252485@main
